src/mem/address_space.h

```diff
@@ -0,0 +1,206 @@
+#include "../pal/pal.h"
+#include "ds/address.h"
+#include "ds/flaglock.h"
+
+#include <array>
+#include <iostream>
+namespace snmalloc
+{
+  template<typename Pal>
+  class AddressSpaceManager : Pal
+  {
+    /**
+     * Implements a power of two allocator, where all blocks are aligned to the
+     * same power of two as their size. This is what snmalloc uses to get
+     * alignment of very large sizeclasses.
+     *
+     * It cannot unreserve memory, so this does not require the
+     * usual complexity of a buddy allocator.
+     */
+
+    // There are a maximum of two blocks for any size/align in a range.
+    // One before the point of maximum alignment, and one after.
+    // As we can add multiple ranges the second entry's block may contain
+    // a pointer to subsequent blocks.
+    std::array<std::array<void*, 2>, bits::BITS> ranges = {};
+
+    // This is infrequently used code, a spin lock simplifies the code
+    // considerably, and should never be on the fast path.
+    std::atomic_flag spin_lock = ATOMIC_FLAG_INIT;
+
+    inline void check_block(void* base, size_t align_bits)
```

Comment:
Please add doc comments for these methods.

Reply:
I disagree with writing doc comments for most of these methods as the name is as descriptive as the comment I would write.

Reply:
I had to read the implementations of these to understand why they are used.

Reply:
Code doesn't normally document "why they are used" but what they do.  Personally, I find the name as descriptive as the comment.  If it has to document its use site, then in this instance it is okay, as most are used in a single place, but normally not.  I could inline them, but the code becomes much less readable.